xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Preconditions;
+import io.grpc.Attributes;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.NameResolver;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.GrpcAttributes;
+import io.grpc.internal.JsonParser;
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.internal.SharedResourceHolder.Resource;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.Executor;
+
+/**
+ * A {@link NameResolver} for resolving gRPC target names with "xds" scheme.
+ *
+ * <p>The implementation is for load balancing alpha release only. No actual VHDS is involved. It
+ * always returns a hard-coded service config that selects the xds_experimental LB policy with
+ * round-robin as the child policy.
+ *
+ * @see XdsNameResolverProvider
+ */
+public final class XdsNameResolver extends NameResolver {
+
+  private static final String SERVICE_CONFIG_HARDCODED = "{"
+          + "\"loadBalancingConfig\": ["
+          + "{\"xds_experimental\" : {"
+          + "\"childPolicy\" : [{\"round_robin\" : {}}]"
+          + "}}"
+          + "]}";
+
+  private final String authority;
+  private final Resource<Executor> executorResource;
+  private final SynchronizationContext syncContext;
+
+  private boolean resolving;
+  private boolean shutdown;
+  private Executor executor;
+  private Listener2 listener;
+
+  XdsNameResolver(String name, Args args, Resource<Executor> executorResource) {
+    URI nameUri = URI.create("//" + checkNotNull(name, "name"));
+    Preconditions.checkArgument(nameUri.getHost() != null, "Invalid hostname: %s", name);
+    authority =
+        Preconditions.checkNotNull(
+            nameUri.getAuthority(), "nameUri (%s) doesn't have an authority", nameUri);
+    this.executorResource = Preconditions.checkNotNull(executorResource, "executorResource");
+    this.syncContext = Preconditions.checkNotNull(args.getSynchronizationContext(), "syncContext");
+  }
+
+  @Override
+  public String getServiceAuthority() {
+    return authority;
+  }
+
+  @Override
+  public void start(final Listener2 listener) {
+    this.listener = listener;
+    executor = SharedResourceHolder.get(executorResource);
+    resolve();
+  }
+
+  @Override
+  public void shutdown() {
+    if (shutdown) {
+      return;
+    }
+    shutdown = true;
+    if (executor != null) {
+      executor = SharedResourceHolder.release(executorResource, executor);
+    }
+  }
+
+  @Override
+  public void refresh() {
+    Preconditions.checkState(listener != null, "not started");
+    resolve();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void resolve() {
+    if (resolving || shutdown) {
+      return;
+    }
+    resolving = true;
+    executor.execute(
```

Comment:
There is no benefit in running this from the executor.

Reply:
Done.